sys/kern/signal.c

```diff
@@ -201,6 +204,93 @@ int do_sigsuspend(proc_t *p, const sigset_t *mask) {
   return ERESTARTNOHAND;
 }
 
+int do_sigtimedwait(proc_t *p, sigset_t waitset, ksiginfo_t *ksi,
+                    timespec_t *tsp) {
+  int error = 0, timeout = 0;
+  sigset_t saved_mask;
+  signo_t sig;
+  thread_t *td = p->p_thread;
+
+  /*
+   * Calculate timeout based on tsp.
+   *
+   * NULL pointer means an infinite timeout.
+   * {.tv_sec = 0, .tv_nsec = 0} means to not block.
+   * Timeout is valid only if 0 <= tv_nsec <= 1e9. If this is not true,
+   * immediately return EINVAL. If tv_sec < 0 immediately return EAGAIN.
+   *
+   * We set the timeout value such that:
+   *   - timeout == -1: do not block
+   *   - timeout == 0: wait indefinitely
```

Comment:
Is it possible? if `timeout` is set to 0 then you change it *always* to -1 or 1.

Reply:
Yes, `timeout == 0` iff `tsp == NULL`

Reply:
ack